packages/server/graphql/traceGraphQL.ts

```diff
@@ -0,0 +1,357 @@
+        } catch (error) {
+          span.setTag('error', error)
+          throw error
+        } finally {
+          finishResolvers(context)
+        }
+      }
+    )
+  }
+  wrappedQuery._datadog_patched = true
+  compiledQuery.query = wrappedQuery
+}
+
+function finishResolvers(contextValue) {
+  const fields = contextValue._datadog_graphql.fields
+
+  Object.keys(fields)
+    .reverse()
+    .forEach((key) => {
+      const field = fields[key]
+
+      if (field.error) {
+        field.span.setTag('error', field.error)
+      }
+      field.span.finish(field.finishTime)
+    })
+}
+
+/**
+ * Wrap all query and mutation resolvers with tracing functions.
+ * This doesn't do any measuring unless it's used together with {@link wrapCompiledQuery}
+ */
+function wrapSchema(tracer: Tracer, config: Config, schema: GraphQLSchema) {
+  wrapFields(tracer, config, schema.getQueryType())
+  wrapFields(tracer, config, schema.getMutationType())
+}
+
+interface Field {
+  parent?: Field
+  span: Span
+  error?: any
+  finishTime?: number
+}
+
+type PatchedContext = {
+  _datadog_graphql: {
+    span: Span
+    fields: {[name: string]: Field}
+    source?: string
+  }
+}
+
+type PatchedGraphQLObjectType = GraphQLOutputType & PatchedMarker
+
+function wrapFields(tracer: Tracer, config: Config, type?: PatchedGraphQLObjectType | null) {
+  if (!type || typeof (type as GraphQLObjectType)?.getFields !== 'function') {
+    return
+  }
+  const fields = (type as GraphQLObjectType).getFields()
+  if (!fields || type._datadog_patched) {
+    return
+  }
+
+  type._datadog_patched = true
+
+  Object.values(fields).forEach((field) => {
+    wrapFieldResolve(tracer, config, field)
+    wrapFieldType(tracer, config, field)
+  })
+}
+
+function wrapFieldResolve(tracer: Tracer, config: Config, field?: GraphQLField<any, any>) {
+  if (!field || !field.resolve) return
+  field.resolve = wrappedResolve(tracer, config, field.resolve)
+}
+
+function wrapFieldType(tracer: Tracer, config: Config, field?: GraphQLField<any, any>) {
+  if (!field || !field.type) return
+
+  let unwrappedType = field.type
+
+  while ((unwrappedType as any).ofType) {
+    unwrappedType = (unwrappedType as any).ofType
+  }
+
+  wrapFields(tracer, config, unwrappedType)
+}
+
+type PatchedGraphQLFieldResolver = GraphQLFieldResolver<any, any> & PatchedMarker
+function wrappedResolve(
+  tracer: Tracer,
+  config: Config,
+  resolve: PatchedGraphQLFieldResolver
+): PatchedGraphQLFieldResolver {
+  if (resolve._datadog_patched || typeof resolve !== 'function') return resolve
+
+  const responsePathAsArray = config.collapse ? withCollapse(pathToArray) : pathToArray
+
+  async function resolveWithTrace(source, args, context, info) {
+    if (!context._datadog_graphql) return resolve(source, args, context, info)
+
+    const path = responsePathAsArray(info && info.path)
+
+    if (config.depth !== undefined && config.depth >= 0) {
+      const depth = path.filter((item) => typeof item === 'string').length
```

Comment:
Right, that's from dd-trace. I think depth only increases for fields of GraphQLObjectType because those can have separate resolve functions. Having lists of lists doesn't increase the depth of those resolvers.